Introducing Hibernate Search Sort DSL

yrodiere opened the pull request:
http://staging.in.relation.to/2016/09/30/introducing-hibernate-search-sort-dsl/

DavideD commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+
+== Overview
+
+The API works pretty much like the query builder. The query builder is the starting point, actually: a new `sort()` method has been added to the `QueryBuilder` interface. Here's what it looks like:
```
I would just keep the sentence:

> A new `sort()` method has been added to the `QueryBuilder` interface:

Everything else seems redundant.

Additional point if you can avoid the passive form, something like the QueryBuilder interface now has an additional sort() method ...`

yrodiere replied:
Done, thanks.
About the passive form, is it because it's bad English or for another reason? Just curious.

DavideD replied:
Active form is easier to read and to understand. Usually, sentences become shorter and the subject of the paragraph become clearer.

It's more of a general guideline, it isn't necessarily bad form. I've found this article and it seems to have good points about when to use it: https://wordcounter.net/blog/2016/09/14/102274_how-to-identify-passive-voice.html 

I'm sure other people will feel differently about it.

DavideD commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+    .createSort()
+----
+ 
+== Advanced use
```
Instead of collect the following features in a unique section, why not to put them in their own section?

yrodiere replied:
The idea was that the "advanced use" section is mainly here to show there's more, not to explain everything else. The reference documentation will be there for people who want extensive information.

But if we want to give the native sort more visibility, then yes, there's not point in keeping that section labelled as "Advanced use".

I changed that.

DavideD replied:
I prefer to keep them separated because it makes section shorter and from the title you can have an overview of the new features.

DavideD commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+  .createSort()
+----
+
+And of course, if you're using Elasticsearch, you may want to take advantage of that brand-new feature that appeared in the last snapshot, the one you just spotted this morning and that would really save you of a lot of trouble on your project. But, hey, the Hibernate Search team is not on the same time zone, and even if they're providing fast support, you're not getting the feature pushed into H ...
```
I would put this in its own section and remove the initial part: `And of course, if you're using Elasticsearch,`

DavideD replied:
I think it's a very good point and deserve to be highlighted more

yrodiere replied:
Right, done.

gsmet commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+
+Of course, one could point out that it's not really backend-independent. The API itself mostly is, but the returned type (`Sort`) is clearly bound to Apache Lucene.
+
+Well, one day at a time: the API in its current form *can* be adapted the be completely agnostic, so it's paving the way to http://hibernate.org/search/roadmap/#hibernate-search-6-x[Hibernate Search 6.x], while still requiring no change to any other contract such as `FullTextQuery.setSort(Sort)`. And that means it will available directly in 5.6.0.Beta3!
```
can be adapted the agnostic -> can be adapted _to_ be _backend-_agnostic?

yrodiere replied:
Right, fixed.

gsmet commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+
+Well, one day at a time: the API in its current form *can* be adapted the be completely agnostic, so it's paving the way to http://hibernate.org/search/roadmap/#hibernate-search-6-x[Hibernate Search 6.x], while still requiring no change to any other contract such as `FullTextQuery.setSort(Sort)`. And that means it will available directly in 5.6.0.Beta3!
+
+So be sure to check it out, and feel free to contact us for any question, problem or simply for feedback!
```
for feedback -> to give us your feedback?

yrodiere replied:
Yeah, not sure it's correct English, but it felt more natural... You're right, better be safe.

gsmet commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+[source, JAVA]
+----
+  .sort()
+    .byNative("authors.name", "{'order':'asc', 'mode': 'min'}")
```
@Sanne @yrodiere we should decide if we choose the 2 parameters form or the one parameter form for byNative before publishing the blog post.

DavideD replied:
Maybe we can mention that we are still deciding between to options and this might change before the final release

yrodiere replied:
@gsmet @DavideD I was expecting us to merge https://github.com/hibernate/hibernate-search/pull/1139 _before_ I publish this blog post anyway. I will fix this snippet before publishing if necessary.
I wouldn't want to confuse user by advertising a feature that does not even has its code released on a branch of the official repo.

DavideD commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+
+That way, if there are two books with the same title in your index, they will always keep the same relative order from one query to another.
+
+The score sort is also available with `byScore()`. Obviously, there's one key difference with that one: the sort is descending by default. Getting the least relevant entities first wouldn't be very useful.
```
Instead of point out the obvious I would explain how you can change the order of the score..
Something like:
`If you need the least relevant results, fear not, we got you covered,  with the byScore().desc()` method.`

yrodiere replied:
Well, actually it's `byScore.asc()`. By ascending score, so the lesser scores first. But yes, I changed the sentence to get to the point (while still stating the obvious, because it might not be so obvious depending on your point of view).

DavideD commented:
@yrodiere Great post.
Overall, I think it would be nicer to create separate sections for the most common cases, spatial and native; instead of a more general `overview` section.

My 2 cents, feel free to pick what you like.

gsmet commented on posts/Yoann/2016-09-30-introducing-hibernate-search-sort-dsl.adoc:
```diff
@@ -0,0 +1,88 @@
+  .createSort()
+----
+
+And of course, if you're using Elasticsearch, you may want to take advantage of that brand-new feature that appeared in the last snapshot, the one you just spotted this morning and that would really save you of a lot of trouble on your project. But, hey, the Hibernate Search team is not on the same time zone, and even if they're providing fast support, you're not getting the feature pushed into H ...
```
When I first read that, it wasn't clear to me that the "feature appeared in the last snapshot" was a feature of Elasticsearch and not a feature of Hibernate Search (which doesn't make sense at all). Maybe "in its last snapshot" would be better?

yrodiere replied:
I changed that to also hint at the fact that this could apply to the Lucene backend, too. I'm not covering this more extensively, since it's less likely to be of use with the Lucene backend. I also addressed the possible confusion in the section title, just in case.
Does it seem clearer to you?

yrodiere commented:
@DavideD @gsmet Thanks for your feedback, I changed the post to fix what you. Feel free to ping me if it's not satisfying. If it is, I'll push that when https://github.com/hibernate/hibernate-search/pull/1139 has been merged.

About the post layout, I wanted to first show we're addressing the most common cases (`byScore`, `byDistance`, and so on) and then show that we're not providing only basic features, without falling in the trap of a reference documentation in a blog post. That was the reason for the "Advanced use" section. I changed it a bit to highlight the native sort part a bit more.

Sanne commented:
It seems phrased as if you wanted to release it before having the API released. I think we should first release Beta3, and only then (immediately after?) release the dedicated blog. WDYT?

I think we're good to release Beta3 immediately.

yrodiere commented:
@Sanne Releasing it after having the API released seems better. Before releasing beta3, maybe we should merge this simple one, though: https://github.com/hibernate/hibernate-search/pull/1169

Sanne commented:
Ok.

On Mon, 3 Oct 2016, 12:50 Yoann Rodière, notifications@github.com wrote:

> @Sanne https://github.com/Sanne Releasing it after having the API
> released seems better. Before releasing beta3, maybe we should merge this
> simple one, though: hibernate/hibernate-search#1169
> https://github.com/hibernate/hibernate-search/pull/1169
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/hibernate/in.relation.to/pull/67#issuecomment-251087499,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AACmTc1MS4E-9_Eq6Pj2mgwFQfYx6kbtks5qwOwWgaJpZM4KLSTV
> .
